note_ends.lua

```diff
@@ -0,0 +1,79 @@
+    finaleplugin.AuthorURL = "http://carlvine.com"
+    finaleplugin.Copyright = "CC0 https://creativecommons.org/publicdomain/zero/1.0/"
+    finaleplugin.Version = "v0.52"
+    finaleplugin.Date = "2022/06/03"
+    finaleplugin.AdditionalMenuOptions = [[ Note Ends Eighths ]]
+    finaleplugin.AdditionalUndoText = [[    Note Ends Eighths ]]
+    finaleplugin.AdditionalPrefixes = [[    eighth_notes = true ]]
+    finaleplugin.AdditionalDescriptions = [[ Change smaller notes followed by rests into eighth notes ]]
+    finaleplugin.MinJWLuaVersion = 0.62
+    finaleplugin.Notes = [[
+        This plug-in aligns the ends of notes followed by a rest to a specific "duration boundary". 
+        Helps improve readability of music with lots of short notes and rests. 
+        It has 2 modes: Eighth Notes and Quarter Notes.
+]]
+    return "Note Ends Quarters", "Note Ends Quarters", "Change smaller notes followed by rests into quarter notes"
+end
+
+-- default to quarter notes for normal operation
+eighth_notes = eighth_notes or false
+
+function expand_note_ends()
+	local delete_next = false
+    local note_value = eighth_notes and finale.NOTE_8TH or finale.NOTE_QUARTER -- [ 512 / 1024 ]
+    local beat_duration = note_value
+	local measure_number = 0
+	local measure = finale.FCMeasure()
+    local compound_meter = false
+	
+	for entry in eachentrysaved(finenv.Region()) do
+	    if measure_number ~= entry.Measure then -- calculate beat duration for each new measure
+	        measure_number = entry.Measure
+	        measure:Load(measure_number)
+	        local time_sig = measure:GetTimeSignature()
+	        beat_duration = time_sig:CalcLargestBeatDuration()
+            compound_meter = (beat_duration % 3 == 0) 
+	    end
+	    
+		if delete_next then -- last note was expanded
+			entry.Duration = 0 -- so delete this rest
+			delete_next = false -- and start over
+            -- OTHERWISE
+		elseif entry:IsNote() -- this is a note
+                and entry:Next() -- with a following entry
+                and entry:Next():IsRest() -- that is a rest
+                and entry.Duration < note_value -- this note is too short
+            then
+
+            local valid = true -- refine the validity criteria
+            local position_in_beat = entry.MeasurePos % beat_duration
+            local note_boundary = entry.MeasurePos % (note_value / 2)
+            local start_beat = math.floor(entry.MeasurePos / beat_duration)
+
+            if beat_duration < (position_in_beat + note_value) -- not enough room in beat for expanded note
+                or note_boundary ~= 0  -- not falling on an allowed duration boundary
+                then
+                valid = false -- don't expand
+                if not compound_meter and not eighth_notes and note_boundary == 0 and start_beat % 2 == 0 then
+                    valid = true  -- special case for quarters on beats 1 & 3
+                end
+            end
+
+            if valid then
```

Comment:
Just from the code itself, I'm unsure why you'd want to modify a valid entry? Given the context of knowing what the script should do it makes sense. But perhaps the code itself could be a bit clearer.

Reply:
Many thanks for the advice Nick. Corrected script follows.